File: Dart/p004-Layers-and-Objects.dart
import 'dart:math';

Random rng = new Random(0);

void main() {
  List<List<double>> X = [
    [1.0, 2.0, 3.0, 2.5],
    [2.0, 5.0, -1.0, 2.0],
    [-1.5, 2.7, 3.3, -0.8]
  ];

  Layer_Dense layer1 = new Layer_Dense(4, 5);
  Layer_Dense layer2 = new Layer_Dense(5, 2);

  layer1.forward(X);
  layer2.forward(layer1.output);
  print(layer2.output);
}

class Layer_Dense {
  late List<List<double>> output;
  late List<List<double>> weights;
  late List<double> biases;

  Layer_Dense(int n_inputs, int n_neurons) {
    weights = [];
    for (int i = 0; i < n_inputs; i++) {
      List<double> temp = [];
      for (int j = 0; j < n_neurons; j++) {
        temp.add(rng.nextDouble() * 0.1);
      }
      weights.add(temp);
    }
    biases = new List<double>.filled(n_neurons, 0);
    output = new List<List<double>>.filled(n_neurons, []);
  }

  void forward(List<List<double>> inputs) {
    output = inputs.dot(weights).addList(biases);
  }
}

/**
 * Everything below is implementation of numpy in dart
 */

extension NNFSV on List<double> {
  // dot product of 2 1D array
  num dot(List list) {
    if (this.length != list.length) {
      throw new Exception("Lists must be the same size");
    } else {
      num result = 0;
      for (int i = 0; i < this.length; i++) {
        result += this[i] * list[i];
      }
      return result;
    }
  }

  // add 1D array to another
  List addList(List list) {
    if (this.length != list.length) {
      throw new Exception("Lists must be the same length");
    }
    return zip({this, list}).map((e) => e[0] + e[1]).toList();
  }
}

// extension to allow numpy functions on 2D list
extension NNFSM on List<List<double>> {
  // dot product of 2 2D arrays
  List<List<double>> dot(List<List<double>> list) {
    var result = new List<double>.filled(this.length, 0)
        .map((e) => List<double>.filled(list[0].length, 0))
        .toList();
    for (var r = 0; r < this.length; r++) {
      for (var c = 0; c < list[0].length; c++) {
        for (var i = 0; i < this[0].length; i++) {
          result[r][c] += this[r][i] * list[i][c];
        }
      }
    }
    return result;
  }

  // transpose 2D array
  List T() {
    List<List<double>> out = [];
    var x = this.length;
    var y = this[0].length;
    var N = x > y ? x : y;
    var M = y < x ? y : x;
    for (int n = 0; n < N; n++) {
      for (int m = 0; m < M; m++) {
        var a = x > y ? n : m;
        var b = y < x ? m : n;
        if (out.length <= b) {
          out.add([]);
        }
        out[b].add(this[a][b]);
      }
    }
    return out;
  }

  // add 1D array to 2D array
  List<List<double>> addList(List<double> list) {
    if (this[0].length != list.length) {
      throw new Exception("List indicies must have same length");
    }
    return this
        .map((e) => zip({e, list}).map((j) => j[0] + j[1]).toList())
        .toList();
  }
}

// for zipping two arrays like python, from:
// https://pub.dev/documentation/quiver/latest/quiver.iterables/zip.html
Iterable<List<T>> zip<T>(Iterable<Iterable<T>> iterables) sync* {
  if (iterables.isEmpty) return;
  final iterators = iterables.map((e) => e.iterator).toList(growable: false);
  while (iterators.every((e) => e.moveNext())) {
    yield iterators.map((e) => e.current).toList(growable: false);
  }
}

File: Kotlin/p004-Layers-and-Objects.kt
import kotlin.random.Random

val rng: Random = Random(0) 

fun main() {
    var X: MutableList<MutableList<Double>> =
            mutableListOf(
                    mutableListOf(1.0, 2.0, 3.0, 2.5),
                    mutableListOf(2.0, 5.0, -1.0, 2.0),
                    mutableListOf(-1.5, 2.7, 3.3, -0.8)
            )

    var layer1: Layer_Dense = Layer_Dense(4, 5)
    var layer2: Layer_Dense = Layer_Dense(5, 2)

    layer1.forward(X)
    layer2.forward(layer1.output)
    println(layer2.output)
}

private class Layer_Dense {
    var output: MutableList<MutableList<Double>> = mutableListOf()
    var weights: MutableList<MutableList<Double>> = mutableListOf()
    var biases: MutableList<Double> = mutableListOf()

    constructor(n_inputs: Int, n_neurons: Int) {
        weights = MutableList(n_inputs) { MutableList(n_neurons) { rng.nextDouble(from = -0.5, until = 1.0) * 0.1 }} // -0.5 to make # more likely to be positive
        biases = MutableList(n_neurons) { 0.0 }
        output = MutableList(n_neurons) { mutableListOf() }
    }

    fun forward(inputs: MutableList<MutableList<Double>>) {
        output = inputs.dot2D(weights).addList2D(biases)
    }
}

/** Everything below is implementation of numpy in kotlin ------------------------ */

// dot product of 2 1D arrays
private fun MutableList<Double>.dot1D(list: MutableList<Double>): Double {
    if (list.size != this.size) {
        throw Exception("Lists must be the same size")
    }
    var result: Double = 0.0
    for (i in 0..this.size - 1) {
        result += this[i] * list[i]
    }
    return result
}

// add 1D array to another 1A array
private fun MutableList<Double>.addList1D(list: MutableList<Double>): List<Double> {
    if (this.size != list.size) {
        throw Exception("Lists must be the same size")
    }
    return this.zip(list) { a, b -> a + b }
}

// dot product of 2 2D arrays
private fun MutableList<MutableList<Double>>.dot2D(
        list: MutableList<MutableList<Double>>
): MutableList<MutableList<Double>> {
    var result: MutableList<MutableList<Double>> =
            MutableList(this.size) { MutableList(list[0].size) { 0.0 } }

    for (r in 0..this.size - 1) {
        for (c in 0..list[0].size - 1) {
            for (i in 0..this[0].size - 1) {
                result[r][c] += (this[r][i] * list[i][c])
            }
        }
    }
    return result
}

// transposes a 2D array
private fun MutableList<MutableList<Double>>.T(): MutableList<MutableList<Double>> {
    var out: MutableList<MutableList<Double>> = mutableListOf()

    var x = this.size
    var y = this[0].size
    var N = if (x > y) x else y
    var M = if (y < x) y else x
    for (n in 0..N - 1) {
        for (m in 0..M - 1) {
            var a = if (x > y) n else m
            var b = if (y < x) m else n
            if (out.size <= b) {
                out.add(mutableListOf())
            }
            out[b].add(this[a][b])
        }
    }
    return out
}

// adds a 1D array to a 2D array
private fun MutableList<MutableList<Double>>.addList2D(
        list: MutableList<Double>
): MutableList<MutableList<Double>> {
    if (this[0].size != list.size) {
        throw Exception("List indicies must be the same size")
    }
    return this.map { a -> a.zip(list) { b, c -> b + c }.toMutableList() }.toMutableList()
}
